integration/tests/cook/test_multi_user.py

```diff
@@ -252,6 +253,64 @@ def test_rate_limit_while_creating_job(self):
             finally:
                 util.kill_jobs(self.cook_url,jobs_to_kill)
 
+    def test_rate_limit_launching_jobs(self):
+        settings = util.settings(self.cook_url)
+        if settings['rate-limit']['job-launch'] is None:
+            pytest.skip()
+
+        # Allow an environmental variable override.
+        user = None
+        name = os.getenv('COOK_LAUNCH_RATE_LIMIT_NAME')
+        if name is not None:
+            user = self.user_factory.user_class(name)
+        else:
+            user = self.user_factory.new_user()
+
+        bucket_size = settings['rate-limit']['job-launch']['bucket-size']
+        token_rate = settings['rate-limit']['job-launch']['tokens-replenished-per-minute']
+        if token_rate < 5 or token_rate > 20:
+            pytest.skip()
+        if bucket_size < 10 or bucket_size > 20:
+            pytest.skip()
+        with user:
+            job_uuids = []
+            try:
+                jobspec = {"command": "sleep 240", 'cpus': 0.03, 'mem': 32}
+
+                self.logger.info(f'Submitting initial batch of {bucket_size+1} jobs')
+                jobs1a, resp1a = util.submit_jobs(self.cook_url, jobspec, bucket_size + 1)
+                job_uuids.extend(jobs1a)
+                self.assertEqual(201, resp1a.status_code, msg=resp1a.content)
+
+                def submit_jobs():
+                    self.logger.info(f'Submitting subsequent batch of {bucket_size+1} jobs')
+                    subsequent_jobs, subsequent_resp = util.submit_jobs(self.cook_url, jobspec, bucket_size + 1)
+                    job_uuids.extend(subsequent_jobs)
+                    self.assertEqual(201, subsequent_resp.status_code, msg=subsequent_resp.content)
+
+                def is_rate_limit_triggered(_):
+                    jobs = util.query_jobs(self.cook_url, True, uuid=job_uuids).json()
+                    waiting_jobs = [j for j in jobs if j['status'] == 'waiting']
+                    self.logger.info(f'There are {len(waiting_jobs)} waiting jobs')
+                    if len(waiting_jobs) > 0:
+                        unscheduled, _ = util.unscheduled_jobs(self.cook_url, *[j['uuid'] for j in waiting_jobs])
+                        self.logger.debug(f'Unscheduled: {json.dumps(unscheduled, indent=2)}')
```

Comment:
I think this log is actually too spammy. It's resulting in pages and pages of output.

Reply:
The really noisy part is when it prints the list of unscheduled jobs. I nuked that line. Just the counts are good for making sure the test is behaving as expected.